browser/brave_ads/brave_stats_updater_helper.cc

```diff
@@ -0,0 +1,132 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_ads/brave_stats_updater_helper.h"
+
+#include "base/files/file_path.h"
+#include "base/functional/bind.h"
+#include "brave/components/brave_ads/common/pref_names.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/common/pref_names.h"
+#include "components/pref_registry/pref_registry_syncable.h"
+
+namespace brave_ads {
+
+BraveStatsUpdaterHelper::BraveStatsUpdaterHelper()
+    : local_state_(g_browser_process->local_state()),
+      profile_manager_(g_browser_process->profile_manager()) {
+#if !BUILDFLAG(IS_ANDROID)
+  last_used_profile_pref_change_registrar_.Init(local_state_);
+  last_used_profile_pref_change_registrar_.Add(
+      ::prefs::kProfileLastUsed,
+      base::BindRepeating(&BraveStatsUpdaterHelper::OnLastUsedProfileChanged,
+                          base::Unretained(this)));
+#endif
+
+  profile_manager_observer_.Observe(profile_manager_);
+}
+
+BraveStatsUpdaterHelper::~BraveStatsUpdaterHelper() {
+  if (current_profile_) {
+    current_profile_->RemoveObserver(this);
+  }
+}
+
+void BraveStatsUpdaterHelper::RegisterLocalStatePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterBooleanPref(ads::prefs::kEnabledForLastProfile, false);
+}
+
+void BraveStatsUpdaterHelper::OnProfileAdded(Profile* profile) {
+#if BUILDFLAG(IS_ANDROID)
+  if (profile == ProfileManager::GetPrimaryUserProfile()) {
+#else
+  base::FilePath last_used_path =
+      local_state_->GetFilePath(::prefs::kProfileLastUsed);
+  if ((!last_used_path.empty() && profile->GetBaseName() == last_used_path) ||
+      (last_used_path.empty() &&
+       profile == ProfileManager::GetPrimaryUserProfile())) {
```

Comment:
We shouldn't call `ProfileManager::GetPrimaryUserProfile`, as it's been deprecated upstream and removed in cr110:

```
  // WARNING: do not use this function on Desktop platforms (Windows, Mac,
  // Linux). See https://crbug.com/1264436 for more info.
  // TODO(https://crbug.com/1264436): restrict this function to Android and
  // ChromeOS.
  //
  // Get the profile for the user which created the current session.
  // Note that in case of a guest account this will return a 'suitable' profile.
  static Profile* GetPrimaryUserProfile();
```

We should probably address this in master and then I can pull the update into cr110 when rebasing next week, but happy to take a different approach too if it makes sense. Could we chat on Monday? cc: @tmancey

Reply:
addressed via #16815